### PR TITLE
MULE-19143: Use InternalEventContext to track foreach context

### PR DESCRIPTION
**Pull request by mlischetti**

### CODE REVIEW

File: core-tests/src/test/java/org/mule/runtime/core/internal/routing/ForeachTestCase.java [hunk 1]
```diff
@@ -87,19 +99,21 @@
 public class ForeachTestCase extends AbstractReactiveProcessorTestCase {
 
   private static final Logger LOGGER = getLogger(ForeachTestCase.class);
+  private static String ERR_NUMBER_MESSAGES = "Not a correct number of messages processed";
+  private static String ERR_PAYLOAD_TYPE = "Type error on processed payloads";
+  private static String ERR_OUTPUT = "Messages processed incorrectly";
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/internal/routing/ForeachTestCase.java:104]: `final`

File: core-tests/src/test/java/org/mule/runtime/core/internal/util/StringHashCodeCollisionGenerator.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import static java.util.Arrays.asList;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class StringHashCodeCollisionGenerator {
+
+  public static List<String> stringsWithSameHashCode(int desiredRecords) {
+    List<String> strings = new ArrayList<>(asList("Aa", "BB"));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/internal/util/StringHashCodeCollisionGenerator.java:17]: assign `asList` directly, no need to create an additional list

File: core/src/main/java/org/mule/runtime/core/internal/message/InternalEvent.java [hunk 3]
```diff
@@ -67,14 +67,26 @@
...
   /**
    * Sets context related to the SDK
    * @param context an {@link EventInternalContext}
    * @since 4.3.0
    */
   <T extends EventInternalContext> void setSdkInternalContext(EventInternalContext<T> context);
 
+  /**
+   * @return a {@link EventInternalContext} with state from the foreach processor
+   * @since 4.3.1
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/message/InternalEvent.java:79]: 4.4, 4.3.1

File: core/src/main/java/org/mule/runtime/core/internal/message/InternalEvent.java [hunk 4]
```diff
@@ -67,14 +67,26 @@
...
+   */
+  <T extends EventInternalContext> EventInternalContext<T> getForeachInternalContext();
+
+  /**
+   * Sets context related to the foreach processor
+   * @param context an {@link EventInternalContext}
+   * @since 4.3.1
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/message/InternalEvent.java:86]: 4.4, 4.3.1

File: core/src/main/java/org/mule/runtime/core/internal/routing/ForeachInternalContextManager.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+import org.mule.runtime.core.api.event.CoreEvent;
+import org.mule.runtime.core.internal.message.EventInternalContext;
+import org.mule.runtime.core.internal.message.InternalEvent;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public class ForeachInternalContextManager {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/ForeachInternalContextManager.java:19]: reduce visibility to package

File: core/src/main/java/org/mule/runtime/core/internal/routing/ForeachRouter.java [hunk 6]
```diff
@@ -176,14 +173,15 @@ private void completeRouter() {
   }
 
   private CoreEvent eventWithCurrentContextDeleted(CoreEvent event) {
-    Map<String, ForeachContext> foreachContextContainer = foreachContextResolver.getCurrentContextFromEvent(event);
-    foreachContextContainer.remove(event.getContext().getId());
-    return foreachContextResolver.eventWithContext(event, foreachContextContainer);
+    removeContext(event);
+    return quickCopy(event, new HashMap<>(0));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/ForeachRouter.java:177]: what do you need this event copy for?

File: core/src/main/java/org/mule/runtime/core/internal/routing/ForeachRouter.java [hunk 7]
```diff
@@ -176,14 +173,15 @@ private void completeRouter() {
...
   }
 
   private CoreEvent prepareEvent(CoreEvent event, String expression) {
-    CoreEvent responseEvent =
-        builder(event).addVariable(owner.getRootMessageVariableName(), event.getMessage(), MULE_MESSAGE).build();
+    CoreEvent responseEvent = InternalEvent.builder(event)
+        .addVariable(owner.getRootMessageVariableName(), event.getMessage(), MULE_MESSAGE)
+        .internalParameters(new HashMap<>())
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/ForeachRouter.java:183]: what is the purpose of cleaning internalParameters here?

File: tests/performance/src/test/java/org/mule/AbstractBenchmarkAssertionTestCase.java [hunk 8]
```diff
@@ -34,7 +33,7 @@
 public abstract class AbstractBenchmarkAssertionTestCase extends AbstractMuleTestCase {
 
   private static final String ENABLE_PERFORMANCE_TESTS_SYSTEM_PROPERTY = "enablePerformanceTests";
-  private static final String NORM_ALLOCATION_RESULT_KEY = "�gc.alloc.rate.norm";
+  private static final String NORM_ALLOCATION_RESULT_KEY = "�gc.alloc.rate.norm";
```
Comment (elrodro83) [inline tests/performance/src/test/java/org/mule/AbstractBenchmarkAssertionTestCase.java:36]: check the encoding you are using

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Comment by mlischetti** [comment]
--validate

File: core/src/main/java/org/mule/runtime/core/internal/routing/ForeachRouter.java (hunk 7 above)
Comment (mlischetti) [reply core/src/main/java/org/mule/runtime/core/internal/routing/ForeachRouter.java]: That was a mistake, internalParameters should not be cleaned

File: core/src/main/java/org/mule/runtime/core/internal/routing/ForeachRouter.java (hunk 6 above)
Comment (mlischetti) [reply core/src/main/java/org/mule/runtime/core/internal/routing/ForeachRouter.java]: This copy is needed to keep actual implementation. 
ForeachRouter creates an EventQuickCopyContextDecorator event instance to process items and returns an EventQuickCopyContextDecorator event instance when foreach ends.
quickCopy implementation is fixing a corner case when a huge number of java.util.Collections.unmodifiableMap are chained causing a StackOverFlow. 

https://github.com/mulesoft/mule/blob/887b71b617472914df85164373e93747ece8203d/core/src/main/java/org/mule/runtime/core/internal/event/DefaultEventBuilder.java#L592

File: tests/performance/src/test/java/org/mule/AbstractBenchmarkAssertionTestCase.java [hunk 9]
```diff
@@ -33,7 +33,7 @@
 public abstract class AbstractBenchmarkAssertionTestCase extends AbstractMuleTestCase {
 
   private static final String ENABLE_PERFORMANCE_TESTS_SYSTEM_PROPERTY = "enablePerformanceTests";
-  private static final String NORM_ALLOCATION_RESULT_KEY = "�gc.alloc.rate.norm";
+  private static final String NORM_ALLOCATION_RESULT_KEY = "gc.alloc.rate.norm";
```
Comment (elrodro83) [inline tests/performance/src/test/java/org/mule/AbstractBenchmarkAssertionTestCase.java:36]: you removed the dot now...

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Review by elrodro83 — APPROVED** [review]

**Pull request merged**